### PR TITLE
Added ability to include directories as an extension of the config file

### LINKED ISSUE #405: Include directive in topgrade.toml
## I want to suggest some general feature

I want to suggest that topgrade add an include directive in its configuration file, so that configuration can be included from another named file. This feature is supported by other tools, including Bash (`source`), tmux (`source-file`), Vim (`source`), OpenSSH (`Include`), Git (`[include]`) and others.

I find this ability helpful when I synchronize dotfiles between computers, but want one small piece of configuration to be different on a particular computer for some reason. For example, my Git configuration file (_.config/git/config_) uses my personal email address, but also includes this section at the bottom:

```gitconfig
[include]
  # Include local configurations, if any.
  path = config.local
```

On my work computer, the _config.local_ file sets the email address to my work email address.

With topgrade, I would do something similar, using this ability to run certain custom commands on some machines, but not others.

I would be happy to help test this.

### PR DESCRIPTION
**Pull request by PolpOnline**
## Standards checklist:

- [x] The PR title is descriptive.
- [x] The code compiles (`cargo build`)
- [x] The code passes rustfmt (`cargo fmt`)
- [x] The code passes clippy (`cargo clippy`)
- [x] The code passes tests (`cargo test`)
- [x] *Optional:* I have tested the code myself
    - [x] I also tested that Topgrade skips the step where needed

I have tested it by myself, but I am not sure if I set all the merge behaviours correctly. 
@openjck you may want to test this.

closes #405

### CODE REVIEW

File: src/config.rs
Comment (Dreaming-Codes) [inline src/config.rs:516]: Wouldn't it be better to simply ignore the file instead of returning an error for the entire read config function?

File: src/config.rs
Comment (Dreaming-Codes) [inline src/config.rs:521]: Same as above

File: src/config.rs
Comment (Dreaming-Codes) [inline src/config.rs:530]: Same as above

File: config.example.toml
Comment (Dreaming-Codes) [inline config.example.toml:4]: Couldn't you use the glob crate to match an entire directory and create a directory dedicated to topgrade configurations, such as `/etc/topgrade.d/`?

**Review by Dreaming-Codes — COMMENTED** [review]
Overall, this is a generally good PR, but I would recommend a few minor improvements.

File: config.example.toml
Comment (PolpOnline) [reply config.example.toml:4]: The linked issue to this pr asked to include files that are in this specific array, whether the directory they are located.

However I could match any file in `$XDG_CONFIG_HOME/topgrade/*` by default, but I would have to ask maintainers if that's a good idea

**Comment by PolpOnline** [comment]
Blocking this from getting merged because the commands and pre_commands field don't have a merge strategy yet, will surely implement it later

File: config.example.toml
Comment (openjck) [inline config.example.toml:3]: Thanks for your interest in this issue.

This part caught my attention. If consistency with other tools is important, it might be good to look at how other tools handle this. It's my understanding that Git, Vim, Bash, and others don't have the main configuration file take precedence. Instead, settings at the bottom of those configuration files take precedence over settings at the tops of those files, and included files are treated as though their contents are inlined.

For example, with the following setup, the effective email address would be _included@example.com_:

```gitconfig
// .gitconfig
[user]
  email = main@example.com

[include]
  path = .gitconfig.included
```

```gitconfig
// .gitconfig.included
[user]
  email = included@example.com
```

However, if the include is moved up, the effective email address would be _main@example.com_:

```gitconfig
// .gitconfig
[include]
  path = .gitconfig.included

[user]
  email = main@example.com
```

```gitconfig
// .gitconfig.included
[user]
  email = included@example.com
```

As far as I know, the same principle is true for Bash, Vim, and others. Settings at the bottom override settings at the top, and included files are treated as though their contents are right there.

Comment (PolpOnline) [reply config.example.toml:3]: This can't be done, not without a breaking change in the way the misc config entries are handled. 
This is because as of now these entries are handled in the global scope, and adding an `[include]` section in between these breaks the following ones, because there is currently no way to specify a return to global scope in TOML files.  

The simplest way to fix this is to add a section named `[misc]` and move all the currently-in-global-scope entries to it.
The problem with this approach is that it breaks all the previously-created config files.

Comment (PolpOnline) [reply config.example.toml:3]: @DottoDev since you are the maintainer of this repository, I'm asking you to tell me how to proceed.

If we are going to migrate all currently-in-global-scope entries to the new `[misc]` section, I'm available to adapt `topgrade` to auto migrate the config file.

File: config.example.toml
Comment (DottoDev) [reply config.example.toml:4]: Matching any file would be cool

File: config.example.toml
Comment (PolpOnline) [reply config.example.toml:3]: I've just pushed a commit to resolve this. I've already tested it myself, but you might want to check it out too @openjck.

**Comment by PolpOnline** [comment]
I think the pr might be ready to be merged if you approve it.

File: config.example.toml
Comment (PolpOnline) [reply config.example.toml:4]: > Matching any file would be cool

I will push a commit to do that

Comment (PolpOnline) [reply config.example.toml]: I think putting them at the beginning of the file (lowest priority) should be fine.

Comment (PolpOnline) [reply config.example.toml]: Solved in [33d1fc6](https://github.com/topgrade-rs/topgrade/pull/421/commits/33d1fc6e6ea10bdd3ba53b06c9767720f8f95a33)

**Review by DottoDev — APPROVED** [review]
lgtm

**Review by DottoDev — CHANGES_REQUESTED** [review]
Cargo fmt and clippy checks need to passbefore a merge can occure.

**Comment by DottoDev** [comment]
Thanks for the work

**Comment by niStee** [comment]
#581 @SteveLauC

**Pull request merged**
